### ISSUE TITLE
Invalid value for "inputMap" parameter: argument must not be null.

### ISSUE THREAD

**Issue by dj-wasabi** [matti/terraform-shell-resource#29]
Hi,

I have the following code:
```
module "files" {
  source  = "matti/resource/shell"
  command = "python ${path.module}/scripts/initialize_vault.py"
  environment = {
    EKS_CLUSTER = var.aws_eks_cluster_eks_cluster.endpoint
    EKS_TOKEN = var.data_aws_eks_cluster_eks_cluster.token
    EKS_CA_CERTIFICATE = base64decode(var.aws_eks_cluster_eks_cluster.certificate_authority.0.data)
  }
}
```

It results in the following error:
```
Error: Invalid function argument

  on .terraform/modules/configuration.files/terraform-shell-resource-1.0.7/main.tf line 104, in resource "null_resource" "contents":
 104:     stdout     = fileexists(local.stdout) ? chomp(file(local.stdout)) : lookup(null_resource.contents_if_missing.triggers, "stdout", "")
    |----------------
    | null_resource.contents_if_missing.triggers is null

Invalid value for "inputMap" parameter: argument must not be null.


Error: Invalid function argument

  on .terraform/modules/configuration.files/terraform-shell-resource-1.0.7/main.tf line 105, in resource "null_resource" "contents":
 105:     stderr     = fileexists(local.stderr) ? chomp(file(local.stderr)) : lookup(null_resource.contents_if_missing.triggers, "stderr", "")
    |----------------
    | null_resource.contents_if_missing.triggers is null

Invalid value for "inputMap" parameter: argument must not be null.


Error: Invalid function argument

  on .terraform/modules/configuration.files/terraform-shell-resource-1.0.7/main.tf line 106, in resource "null_resource" "contents":
 106:     exitstatus = fileexists(local.exitstatus) ? chomp(file(local.exitstatus)) : lookup(null_resource.contents_if_missing.triggers, "exitstatus", -1)
    |----------------
    | null_resource.contents_if_missing.triggers is null

Invalid value for "inputMap" parameter: argument must not be null.
```

And I don't know why?

Running on a Mac:
```
Terraform v0.12.24
+ provider.aws v2.65.0
+ provider.helm v1.2.2
+ provider.kubernetes v1.11.3
+ provider.null v2.1.2
+ provider.random v2.2.1
+ provider.template v2.1.2
```

**Comment by jongio** [comment]
I'm seeing the same with
```
module "files" {
  source  = "matti/resource/shell"
  command = "dir"
}

output "my_files" {
  value = module.files.stdout
}
```

**Comment by channaj** [comment]
Getting the same issue in WSL
```Terraform v0.12.29
+ provider.azuread v0.11.0
+ provider.azurerm v2.13.0
+ provider.null v2.1.2
+ provider.random v2.3.0
+ provider.time v0.5.0

**Comment by channaj** [comment]
Have you managed to find the problem @dj-wasabi ?

**Comment by dj-wasabi** [comment]
Hi @channaj 

No, stopped using this module.

**Comment by jongio** [comment]
I don't need to get values back from the command, so I just created a module to execute a script like this:
https://github.com/jongio/azsdkdemo/tree/master/iac/terraform/modules/script



And then in my main file:
https://github.com/jongio/azsdkdemo/blob/master/iac/terraform/main.tf#L136

```
module "script" {
  source = "./modules/script"
  script = "./tweaks.sh"
  environment = {
    TEXT_ANALYTICS_NAME  = azurerm_cognitive_account.text_analytics.name,
    FORM_RECOGNIZER_NAME = azurerm_cognitive_account.form_recognizer.name,
    RESOURCE_GROUP_NAME  = azurerm_resource_group.rg.name
  }
}
```

https://github.com/jongio/azsdkdemo/blob/master/iac/terraform/tweaks.sh

**Comment by matti** [comment]
the internal implementation of this is such a big hack that I don't know what is going on here.

**Comment by bvarga** [comment]
getting the same issue. 

Could solve it by incrementally setting up the config. 
Just added this module after everything was up to date, then added without trigger, then set up the trigger part of the config.

**Comment by olenm** [comment]
i think this solves it, at least for 0.13.5 - (im in the middle of upgrading 12 to 14); though it likely needs more testing, but my use-case is rather limited at the moment. The below did get me a TF ouput (and my use cases query-a-thing and then use stdout to optionally modify-a-thing, both using this module).


`main.tf` lines 106,107,108
```
    stdout     = fileexists(local.stdout)     ? chomp(file(local.stdout))     : lookup(  (null_resource.contents_if_missing.triggers == null ? {} : null_resource.contents_if_missing.triggers), "stdout", "")
    stderr     = fileexists(local.stderr)     ? chomp(file(local.stderr))     : lookup(  (null_resource.contents_if_missing.triggers == null ? {} : null_resource.contents_if_missing.triggers), "stderr", "")
    exitstatus = fileexists(local.exitstatus) ? chomp(file(local.exitstatus)) : lookup(  (null_resource.contents_if_missing.triggers == null ? {} : null_resource.contents_if_missing.triggers), "exitstatus", -1)
```

Edit: yep, it blows up on apply, TF reporting values of plan/apply yield different data for triggers. I may tinker with this some more, is a fun one 


Edit 2: 
this seems to get the job done for TF 0.13.5 and 0.14.5
```
    stdout     = fileexists(local.stdout)     ? chomp(file(local.stdout))     :  ( null_resource.contents_if_missing.triggers == null ? "" : lookup(null_resource.contents_if_missing.triggers, "stdout", "")      )
    stderr     = fileexists(local.stderr)     ? chomp(file(local.stderr))     :  ( null_resource.contents_if_missing.triggers == null ? "" : lookup(null_resource.contents_if_missing.triggers, "stderr", "")      )
    exitstatus = fileexists(local.exitstatus) ? chomp(file(local.exitstatus)) :  ( null_resource.contents_if_missing.triggers == null ? -1 : lookup(null_resource.contents_if_missing.triggers, "exitstatus", -1)  ) 
```

```
terraform version
Terraform v0.14.5
+ provider registry.terraform.io/hashicorp/aws v2.70.0
+ provider registry.terraform.io/hashicorp/null v3.0.0
+ provider registry.terraform.io/hashicorp/random v3.0.1
```

**Comment by matti** [comment]
huh, I have no idea if that works or not. maybe someone can verify?

**Comment by olenm** [comment]
@matti 
using TF 0.12.12, 0.13.5, 0.14.5 with  `terraform console`, using the following code generates the same errors:
```
> lookup( null, "a" )

Error: Invalid function argument

  on <console-input> line 1:
  (source code not available)

Invalid value for "inputMap" parameter: argument must not be null.
```


what the fix does is checks if the `null_resource.contents_if_missing.triggers` is null before passing it into the function `lookup` and processes it accordingly, if its null it gives the same 'default' value as if the lookup were to not have the correct map within it.


Edit: the PR is building again via travisci; I re-pushed it with the proper issue-tag reference just now so it links properly

**Comment by matti** [comment]
yeah I get that, but this whole thing is sooo messy.

I'll just merge and see what happens hehe.

Btw, I might give a shot at writing a shell provider instead. This module approach is not great.

**Comment by seboudry** [comment]
Hi @matti !
Any chance to have a release with this fix? ;)

**Comment by matti** [comment]
@seboudry released v1.2.0
